MAINT: version bounds 1.9.0rc1

tylerjereddy opened the pull request:
* I never get this quite right, but here is the
first draft of the version bounds for `1.9.0rc1`,
which is probably even more complicated with some
of the meson stuff until I'm used to that

* for reference, last branching we did this:
gh-15191

* I think if NumPy releases before us we can, in theory,
squeeze another +1 version if we're careful to address
warnings (I think we actually ignored some though..)

tylerjereddy commented on pyproject.toml:
```diff
@@ -61,11 +61,11 @@ maintainers = [
 # release branches, see:
 #     https://scipy.github.io/devdocs/dev/core-dev/index.html#version-ranges-for-numpy-and-other-dependencies
-requires-python = ">=3.8"
+requires-python = ">=3.8,<3.11"
```
I tried two different versions and got:

- ` packaging.specifiers.InvalidSpecifier: Invalid specifier: '>=3.8,<3.11'`
- `` packaging.specifiers.InvalidSpecifier: Invalid specifier: '>=3.8,<3.12'``

Unless I'm missing something obvious, an identical format worked just fine on the last release branch: https://github.com/scipy/scipy/blob/maintenance/1.8.x/pyproject.toml#L58

Is this through the meson stuff now and therefore different or?

rgommers replied:
Let's make this `<3.12`, because in 1.9.x we're going to have to support Python 3.11 and people are already testing with it.

The first time we can/should upload wheels is the first bugfix release after 2022-08-01 (RC1 date), but we should allow building from source now.

tylerjereddy replied:
What to do about the traceback though?

rgommers commented on meson.build:
```diff
@@ -6,7 +6,7 @@ project(
   version: '1.9.0.dev0',
   license: 'BSD-3',
-  meson_version: '>= 0.60',
+  meson_version: '>= 0.60,< 0.70.0',
```
This is a little arbitrary, but we should probably set it a little tighter. Latest released version is `0.62.2`, and we need to get a bunch of features that will help into 0.63-0.65. So I'd take a guess saying `<0.66` is about right.

rgommers commented on pyproject.toml:
```diff
@@ -10,13 +10,13 @@
+    "meson-python>=0.5.0,<0.6.0",
+    "Cython>=0.29.21,<3.0",
+    "pybind11>=2.4.3,<2.10.0",
+    "pythran>=0.9.12,<0.12.0",
```
Can we up this to `>=0.9.12,<0.12.0` like I just did in `main`? Clang 13 requires 0.11.0, and I've seen multiple issues about that now.

tylerjereddy replied:
Isn't that what I did already?

rgommers replied:
Sorry, I think I meant `>=0.11.0,<0.12.0`. Doesn't matter much though; the range includes the fix for Clang 13, and users are mostly going to get the latest version anyway. 0.9.12 and 0.10.x are good for everything else. 

Resolving.

rgommers reviewed (commented):
Thanks Tyler. Overall looks good, a few tweks suggested.

stefanv commented:
Can you explain the rationale behind an upper bound on Python? Presumably, people should be able to build on whichever version of Python they like.

I'm still -1 on NumPy pinning, and would prefer a helpful warning. But I know the SciPy team has already converged on this as an acceptable practice.

tylerjereddy commented:
Maybe deprecations in the Python language itself? Not trying to defend it aggressively, but I suppose that *eventually* the probability of a successful build from source drops as the versions/years go by?

stefanv commented:
I suspect that those deprecations are quite rare, and if users run into them we can simply fix them (this is during the build process, after all).

tylerjereddy commented:
I wonder if we'd feel compelled to release new source distributions in that event, which then means we need to rebuild new binary wheels for PyPI with a much newer version of Python than the old release? Which then might mean a new `pip` which has new rules or something? I dunno, I don't feel too strongly about it, but just kind of imagining what could go wrong given what tends to go wrong even for "normal" releases.

If minor shims for developers are fair game though, can't you just delete the version bounds in the source if a little bit of hacking is reasonable anyway?

I guess this way we get "insurance" against maintenance complaints (i.e., this is our policy, hack as you see fit if you want support beyond it).

stefanv commented:
Does this pin prevent future versions of Python from attempting to build the source package? If so, I suppose it makes some sense, since as you say devs can still apply a small patch.

tylerjereddy commented:
> Does this pin prevent future versions of Python from attempting to build the source package?

I tried (`python -m pip install -v .`, which now goes through `meson`) locally in a Python `3.10` `venv` with:

```diff
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -61,7 +61,7 @@ maintainers = [
 # Note: Python and NumPy upper version bounds should be set correctly in
 # release branches, see:
 #     https://scipy.github.io/devdocs/dev/core-dev/index.html#version-ranges-for-numpy-and-other-dependencies
-requires-python = ">=3.8,<3.11"
+requires-python = "<3.10"
 dependencies = [
```

And what happened was the entire source was built successfully, and then at the very last step, you see:

`ERROR: Package 'scipy' requires a different Python: 3.10.4 not in '<3.10'`

and nothing gets installed in site packages, etc. So, I suppose the answer is a (very annoyingly implemented) yes. At least the build is way faster so it is not like waiting 8-10 minutes as it used to be.

That said, this PR can't fly at the moment with the version pin because of the failure in CI, which I can also reproduce locally (see below the fold)--apparently the `pep621` PyPI package can't parse the `'>=3.8,<3.11'`, which we could parse before switching build backends. I suspect that's a bug rather than an ecosystem statement on upper version bounds, though I could be wrong cc @FFY00.

<details>

```
  Running command Getting requirements to build wheel
  Traceback (most recent call last):
    File "/home/treddy/python_virtual_envs/python_310_venv_scipy_general/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
      main()
    File "/home/treddy/python_virtual_envs/python_310_venv_scipy_general/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/treddy/python_virtual_envs/python_310_venv_scipy_general/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/mesonpy/__init__.py", line 805, in get_requires_for_build_wheel
      with _project(config_settings) as project:
    File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
      return next(self.gen)
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/mesonpy/__init__.py", line 784, in _project
      with Project.with_temp_working_dir(
    File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
      return next(self.gen)
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/mesonpy/__init__.py", line 479, in with_temp_working_dir
      yield cls(source_dir, tmpdir, build_dir)
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/mesonpy/__init__.py", line 380, in __init__
      self._metadata = pep621.StandardMetadata.from_pyproject(self._config, self._source_dir)
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/pep621.py", line 227, in from_pyproject
      packaging.specifiers.Specifier(requires_python_string) if requires_python_string else None,
    File "/tmp/pip-build-env-5seiz966/overlay/lib/python3.10/site-packages/packaging/specifiers.py", line 98, in __init__
      raise InvalidSpecifier(f"Invalid specifier: '{spec}'")
  packaging.specifiers.InvalidSpecifier: Invalid specifier: '>=3.8,<3.11'
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/treddy/python_virtual_envs/python_310_venv_scipy_general/bin/python /home/treddy/python_virtual_envs/python_310_venv_scipy_general/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /tmp/tmp7huy0k7n
  cwd: /home/treddy/github_projects/scipy
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
(python_310_venv_scipy_general) treddy@adobe:~/github_projects/scipy$ 

```

</details>

FFY00 commented:
That is likely a bug, I need to look better into the standards though.

Regarding it erroring out at the last step. It seems the wheel is successfully built but the metadata says it is not compatible with Python 3.10. We could verify if the interpreter being used for the build is compatible with the specified `requires-python` field preemptively, as I probably no valid use case would require the backend to succeed in such case.

rgommers commented:
> That is likely a bug, I need to look better into the standards though.

The standard says nothing, and there's a long and inconclusive thread on Discourse. SciPy 1.7.x and 1.8.x do it like this already, and both `setuptools` and `pip` install upper bounds just fine. 

There's folks who would like to not have upper bounds, and advocate instead for raising an exception in `setup.py` dynamically at build time. This is clearly a little strange (remove metadata to attempt a build that you know for certain will raise an exception), but there's complicated reasons for and against. 

> suspect those deprecations are quite rare,

This is not true, we have fixes for changes in new Python versions in every release of both NumPy and SciPy. if you don't believe me, just try to take a recent `1.X.0` release and try to build it with a version of Python that was released more than a year after that particular minor release. I'm not sure if your chance of success is zero or non-zero, but I am sure that it is very low.

> and if users run into them we can simply fix them (this is during the build process, after all).

Wanting an obscure error halfway through the build rather than a "this is not supported" message right at the start is just a poor choice. Users would be much better served by getting an informative error saying "this Python is too new, please use the `main` branch or a newer version of SciPy". 

At this point I am deciding to stop caring, because there's better ways to spend my time than to re-litigate this yet again. So I'll unsubscribe from this thread.

stefanv commented:
> Users would be much better served by getting an informative error saying "this Python is too new, please use the `main` branch or a newer version of SciPy".

I see now that this patch is only applied on the 1.9.x branch, and that the main branch does not have similar pinning—so, that seems fine (as long as users always can always try to build the latest source on latest Python).

stefanv commented:
Thanks for taking the time to think through the Python version pinning with me. After further consideration, I get why the Python pin makes sense, specifically because it affects source builds of a release branch only. The built wheels are already tied to a Python version, and the pin does not preclude anything that isn't otherwise possible (other than trying to build an old SciPy on a newer Python). The latest source build is the sensible place to support new Pythons.

I'm sorry that the conversation frustrated you @rgommers. We're all tuning in to a lot of things simultaneously, so sometimes a reminder of why we do things the way we do is helpful.

FFY00 commented:
> The standard says nothing, and there's a long and inconclusive thread on Discourse. SciPy 1.7.x and 1.8.x do it like this already, and both `setuptools` and `pip` install upper bounds just fine.

I had a look and this field is a pep 440 version specifier, which allows multiple comma separated clauses, including upper bounds. The issue seems to be a bug in `pep621` or `packaging` that does not allow a list.

I am not sure if we are talking about the same thing here. Did I miss something?

tylerjereddy commented:
I think I've addressed the reviewer comments in my latest commit, however at least locally I get a different type of error when trying to build on Linux in a Python 3.9 `venv` now:

`python -m pip install -v .`

```
  Traceback (most recent call last):
    File "/home/tyler/python_3_9_12_scipy_general/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 363, in <module>
      main()
    File "/home/tyler/python_3_9_12_scipy_general/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 345, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/tyler/python_3_9_12_scipy_general/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py", line 130, in get_requires_for_build_wheel
      return hook(config_settings)
    File "/tmp/pip-build-env-_ecfq9dy/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 805, in get_requires_for_build_wheel
      with _project(config_settings) as project:
    File "/home/tyler/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/python-3.9.12-osore4bvxpxd66nnaw6pvjrmkf7csyia/lib/python3.9/contextlib.py", line 119, in __enter__
      return next(self.gen)
    File "/tmp/pip-build-env-_ecfq9dy/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 784, in _project
      with Project.with_temp_working_dir(
    File "/home/tyler/github_projects/spack/opt/spack/linux-ubuntu22.04-skylake_avx512/gcc-11.2.0/python-3.9.12-osore4bvxpxd66nnaw6pvjrmkf7csyia/lib/python3.9/contextlib.py", line 119, in __enter__
      return next(self.gen)
    File "/tmp/pip-build-env-_ecfq9dy/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 479, in with_temp_working_dir
      yield cls(source_dir, tmpdir, build_dir)
    File "/tmp/pip-build-env-_ecfq9dy/overlay/lib/python3.9/site-packages/mesonpy/__init__.py", line 430, in __init__
      self._metadata.dynamic.remove('version')
  ValueError: list.remove(x): x not in list
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  full command: /home/tyler/python_3_9_12_scipy_general/bin/python /home/tyler/python_3_9_12_scipy_general/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py get_requires_for_build_wheel /tmp/tmpotyqonyz
  cwd: /home/tyler/github_projects/scipy
  Getting requirements to build wheel ... error
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

```

tylerjereddy commented:
yeah, CI agrees with me here it seems, @FFY00 thoughts?

tylerjereddy commented:
My initial testing suggests we'll be in better shape when https://github.com/FFY00/meson-python/pull/77 gets merged. I think one source of confusion for me was that the `pep621` update on its own made things look even worse, but when combined with the latest `meson-python` looks solid with this feature branch as well.

h-vetinari commented:
> [...] the `pep621` update [...]

On that topic, `pep621` is now called [`pyproject-metadata`](https://pypi.org/project/pyproject-metadata/), see https://github.com/FFY00/python-pyproject-metadata/issues/10

tylerjereddy commented:
Ok, the release history shows `pep621` active as of two days ago: https://pypi.org/project/pep621/#history

I can't keep track of it all perfectly, just hoping the pieces are in place to do an rc soon.

h-vetinari commented:
> Ok, the release history shows `pep621` active as of two days ago

True, but `0.4.0.post2` is still `< 0.5.0`. ;-)

> I can't keep track of it all perfectly, just hoping the pieces are in place to do an rc soon.

🤞

FFY00 commented:
Yeah, `pep621` was renamed to `pyproject-metadata`, `meson-python` 0.6.0 will have the updated dependency.

tylerjereddy commented:
With the release of `meson-python` `0.6.0` I've locally confirmed that the issues above seem to be resolved on this branch locally.

I rebased/resolved merged conflicts and pushed in--hopefully the CI agrees.
